File: sol/solution/0101-0200/0108/Solution.ex
# Definition for a binary tree node.
#
defmodule TreeNode do
  @type t :: %__MODULE__{
          val: integer,
          left: TreeNode.t() | nil,
          right: TreeNode.t() | nil
        }
  defstruct val: 0, left: nil, right: nil
end
defmodule Solution do
  @spec sorted_array_to_bst(nums :: [integer]) :: TreeNode.t | nil
  def sorted_array_to_bst(nums) do
    build_bst(nums, 0, length(nums) - 1)
  end
  defp build_bst(_nums, left, right) when left > right, do: nil
  defp build_bst(nums, left, right) do
    mid = left + div(right - left, 2)
    %TreeNode{
      val: Enum.at(nums, mid),
      left: build_bst(nums, left, mid - 1),
      right: build_bst(nums, mid + 1, right)
    }
  end
end

File: sol/solution/0101-0200/0108/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-spec sorted_array_to_bst(Nums :: [integer()]) -> #tree_node{} | null.
sorted_array_to_bst(Nums) ->
    build_bst(Nums, 1, length(Nums)).
build_bst(_Nums, Left, Right) when Left > Right ->
    null;
build_bst(Nums, Left, Right) ->
    Mid = Left + (Right - Left) div 2,
    #tree_node{
        val = lists:nth(Mid, Nums),
        left = build_bst(Nums, Left, Mid - 1),
        right = build_bst(Nums, Mid + 1, Right)
    }.